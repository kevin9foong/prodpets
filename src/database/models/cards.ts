import db from '../index';
// import firebase from 'firebase/app';
import 'firebase/firestore';
import { ChecklistItem } from '../../components/Checklist';

export type cardStatus = 'completed' | 'in progress' | 'incomplete';
export interface CardModelWithUid<DateType = Date> extends CardModel<DateType> {
	uid: string
}
export interface CardModel<DateType = Date> { 
    title: string, 
    description: string,
	startdate: DateType,
	duedate: DateType, 
	effortHours: number, 
	tags: string[],
	status: cardStatus, 
	checklistItems: ChecklistItem[]
}

// make sure we use the local offline store as much as possible 
// - unless we explicitly request to sync with online.
const getOptions = {
	source: 'cache'
};

export const saveCard = (userUid: string, card: CardModel) => {
	return db.collection('users').doc(userUid)
		.collection('cards')
		.add({
			...card
		});
};

// TODO: add sorting, filter data in the future.
/**
 * Returns a promise of user items.
 * @param userUid unique user id
 * @returns 
 */
export const fetchCards = async (userUid: string): Promise<CardModelWithUid[]> => {
	console.log('FETCHING TRIGGERED');
	const cardsQuerySnapShot = await db.collection('users').doc(userUid)
		.collection('cards').limit(5).get();
	const cardsQuerySnapShotDocs = cardsQuerySnapShot.docs; 

	const cardsDataPromises = cardsQuerySnapShotDocs.map(async doc => {
		const uid = doc.id; 
		const docData = await doc.data(); 
		return {
			uid, 
			...docData
		};
	});

	// need to store into Redux store. 
	return Promise.all(cardsDataPromises).then(
		(cardsData) => {
			return cardsData.map(card => {
				return ({
					...card,
					startdate: card.startdate.toDate(),
					duedate: card.duedate.toDate()
				} as CardModelWithUid);});});
};


export const updateCard = (userUid: string, cardUid: string, card: CardModel) => {
	return db.collection('users').doc(userUid).collection('cards')
		.doc(cardUid).update({
			...card
		});
};