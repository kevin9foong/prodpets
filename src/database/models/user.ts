import db from '../index';
import 'firebase/firestore';

export type UserModel = { 
    uid: string, 
    displayName: string, 
    email: string, 
    photoURL: string
}

export const saveUser = (user: UserModel): Promise<void> => {
	return db.collection('users').doc(user.uid)
		.set({
			name: user.displayName, 
			email: user.email, 
			photoURL: user.photoURL
		});
};