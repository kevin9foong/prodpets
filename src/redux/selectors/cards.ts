// TODO: add more selectors
export const selectCards = state => state.cards;
export const selectCardByUID = (uid: string) => state => state.cards.filter(card => card.uid === uid)[0]; 
