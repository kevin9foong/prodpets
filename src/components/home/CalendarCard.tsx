import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import { CardModelWithUid } from '../../database/models/cards';
import CalendarCardStyles from '../../styles/components/home/CalendarCard.style';
import { getTimeRange } from '../../util/timeformatter';

type CalendarCardProps = {
  cardInfo: CardModelWithUid;
  onCardClick: (a: CardModelWithUid) => void;
};

const CalendarCard = ({
	cardInfo,
	onCardClick,
}: CalendarCardProps): JSX.Element => {
	return (
		<TouchableOpacity onPress={() => onCardClick(cardInfo)}>
			<View style={CalendarCardStyles.container}>
				<Text style={{ paddingBottom: 5, color: 'white' }}>
					{cardInfo.title}
				</Text>
				<Text style={{ color: 'white' }}>
					{getTimeRange(
						new Date(cardInfo.startdate),
						new Date(cardInfo.duedate)
					)}
				</Text>
			</View>
		</TouchableOpacity>
	);
};

export default CalendarCard;
