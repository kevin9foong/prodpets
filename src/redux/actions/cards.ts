import { CardModel, CardModelWithUid } from '../../database/models/cards';
import { fetchCards as fetchCardsFromDb } from '../../database/models/cards';

export const fetchCards = async (dispatch, getState) => {
	const userUid = getState().user.userUid;
	const cards = !userUid ? [] : await fetchCardsFromDb(userUid);
	dispatch({
		type: 'cards/fetchCards', 
		payload: {
			cards
		}
	});
};

export const addCard = (uid, cardInfo: CardModel) => ({
	type: 'cards/addCard', 
	payload: {
		cardInfo,
		uid
	}
});

export const updateCard = (cardInfo: CardModelWithUid) => ({
	type: 'cards/updateCard', 
	payload: {
		cardInfo
	}
});

export const deleteCard = (cardUid: string) => ({
	type: 'cards/deleteCard', 
	payload: {
		cardUid
	}
});