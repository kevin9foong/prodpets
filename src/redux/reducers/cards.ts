import { CardModelWithUid } from '../../database/models/cards';
import { addCardAction, deleteCardAction, fetchCardsAction, updateCardAction } from '../actions/cards';

export type Cards = {
	// cardUid: cardData 
	[uid: string]: CardModelWithUid
}

const initialState: Cards = {}; 

const cardsReducer = (state = initialState, action: addCardAction | updateCardAction | deleteCardAction | fetchCardsAction): Cards => {
	switch(action.type) {
	case 'cards/fetchCards': { 
		// TODO: fix implementation later with syncing online mode with offline state
		// return [...state];
		return {};
	}
	case 'cards/addCard': {
		return {...state, [action.payload.cardInfo.uid]: action.payload.cardInfo};
	}
	case 'cards/updateCard': {
		const newCards = {...state};
		newCards[action.payload.cardInfo.uid] = action.payload.cardInfo;
		return newCards;
	}
	case 'cards/deleteCard': {
		const newCards = {...state};
		delete newCards[action.payload.uid]; 
		return newCards;
	}
	default: {
		return state;
	}
	}
};


export default cardsReducer; 