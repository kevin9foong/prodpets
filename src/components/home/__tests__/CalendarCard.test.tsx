import CalendarCard from '../CalendarCard';
import { CardModelWithUid } from '../../../database/models/cards';
import { render } from '@testing-library/react-native'; 
import React from 'react';

test('Calendar Card single day event', () => {
  const info = {
    item: {
      title: "dummy Event",
      description: "dummy description",
      startdate: new Date('July 4, 2021 15:24:00'),
      duedate: new Date('July 4, 2021 16:00:00'),
      uid: '',
    }
  };
  const clickHandler = (card: CardModelWithUid) => {};
  const { toJSON } = render(<CalendarCard info={info} clickHandler={clickHandler}/>)
  expect(toJSON()).toMatchSnapshot();
})

test('Calendar Card multi day event display', () => {
  const info = {
    item: {
      title: "dummy Event",
      description: "dummy description",
      startdate: new Date('July 4, 2021 15:24:00'),
      duedate: new Date('July 7, 2021 16:00:00'),
      uid: '',
    }
  };
  const clickHandler = (card: CardModelWithUid) => {};
  const { toJSON } = render(<CalendarCard info={info} clickHandler={clickHandler}/>)
  expect(toJSON()).toMatchSnapshot();
})