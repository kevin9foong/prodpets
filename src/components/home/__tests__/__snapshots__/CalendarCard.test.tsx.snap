// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Card multi day event display 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#84A9FF",
        "borderRadius": 20,
        "flex": 1,
        "marginBottom": 5,
        "marginRight": 30,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "paddingBottom": 5,
        }
      }
    >
      dummy Event
    </Text>
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      4/7 - 7/7
    </Text>
  </View>
</View>
`;

exports[`Calendar Card single day event 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#84A9FF",
        "borderRadius": 20,
        "flex": 1,
        "marginBottom": 5,
        "marginRight": 30,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "paddingBottom": 5,
        }
      }
    >
      dummy Event
    </Text>
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      15:24 - 16:00
    </Text>
  </View>
</View>
`;

exports[`Check Calendar Card click functionality 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#84A9FF",
        "borderRadius": 20,
        "flex": 1,
        "marginBottom": 5,
        "marginRight": 30,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "paddingBottom": 5,
        }
      }
    >
      dummy Event
    </Text>
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      15:24 - 16:00
    </Text>
  </View>
</View>
`;
